Fixes OWASP/www-community#992

jakekarnes42 opened the pull request:
This is a significant rewrite of the current [Insufficient Session-ID Length page](https://owasp.org/www-community/vulnerabilities/Insufficient_Session-ID_Length). 

The motivation for this rewrite is explained in #992. 

I've done my best to be careful with the formatting and linked images, but I didn't have a great way to validate this formatting is working correctly outside of the markdown preview. The images don't display in the preview, but I linked the new images in the same manner as the original page. 

Feedback is welcome!

Fixes OWASP/www-community#992

kingthorin commented:
https://owasp.org/migration/ provides info on setting up a local testing environment.

jakekarnes42 commented:
Thank you for the quick reply. I've used the local environment as suggested to validate the formatting. I made a few improvements and I think the formatting is clear now. Here are two quick examples from some formatting-heavy sections:

### Example 1:
![image](https://github.com/user-attachments/assets/6de27b7f-104a-462f-a43a-300b93bd7b20)

### Example 2:
![image](https://github.com/user-attachments/assets/18241fbb-3183-48b7-907e-df865f04718f)

kingthorin commented on pages/vulnerabilities/Insufficient_Session-ID_Length.md:
```diff
@@ -13,33 +13,124 @@ tags: vulnerability, Insufficient Session-ID Length
+Session identifiers must have **at least 64 bits of entropy** to prevent brute-force session guessing attacks. It's important to understand that session ID length alone is not a sufficient indicator of security. Instead, the critical factor is the entropy (or randomness) contained within the session identifier. While the length of the session ID can affect how much entropy is encoded, the actual  ...
 
-The WebLogic deployment descriptor should specify a session identifier length of at least 128 bits. A shorter session identifier leaves the application open to brute-force session guessing attacks. If an attacker can guess an authenticated user's session identifier, they can take over the user's session. The remainder of this explanation will detail a back-of-the-envelope justification for a 128  ...
+### Key Takeaways:
```
Drop the colon

kingthorin commented on pages/vulnerabilities/Insufficient_Session-ID_Length.md:
```diff
@@ -13,33 +13,124 @@ tags: vulnerability, Insufficient Session-ID Length
 ## Description
 
-Session identifiers should be at least 128 bits long to prevent brute-force session guessing attacks.
+Session identifiers must have **at least 64 bits of entropy** to prevent brute-force session guessing attacks. It's important to understand that session ID length alone is not a sufficient indicator of security. Instead, the critical factor is the entropy (or randomness) contained within the session identifier. While the length of the session ID can affect how much entropy is encoded, the actual  ...
```
Could you ensure consistency in ID vs id vs Id?

This paragraph uses one and two. Chances are there are other places in the doc too.

IMHO full caps is the way to go, but consistency is king.

kingthorin commented on pages/vulnerabilities/Insufficient_Session-ID_Length.md:
```diff
@@ -13,33 +13,124 @@ tags: vulnerability, Insufficient Session-ID Length
+3. **16-byte hexadecimal session ID with fixed characters**:
+    - Example Session ID #1: `DECAFBAD12345678`
+    - Example Session ID #2: `DECAFBAD87654321`
+    - Example Session ID #3: `DECAFBAD12341234`
```
🚫☕️👎 that's wonderful. I appreciate this bit of humour

kingthorin commented on pages/vulnerabilities/Insufficient_Session-ID_Length.md:
```diff
@@ -13,33 +13,124 @@ tags: vulnerability, Insufficient Session-ID Length
+### Example Calculations
+The following provides some example calcuations based on different attack scenarios. 
+
+#### Example 1: 
```
I'd drop all the trails non colons from headings, they seem unnecessary

kingthorin commented on pages/vulnerabilities/Insufficient_Session-ID_Length.md:
```diff
@@ -13,33 +13,124 @@ tags: vulnerability, Insufficient Session-ID Length
 
 ## Risk Factors
 
 - Attackers that are try to obtain a valid session ID for [Session hijacking](../attacks/Session_hijacking_attack).
```
trying

I know that's an existing typo but you may as well fix it

jakekarnes42 commented:
Thank you for the thoughtful feedback! My latest commit addresses those issues along with 2 minor content changes:

1. I added a brief explanation of what is meant by "bits of entropy" at the beginning. The first paragraph now reads:
>Session identifiers must have **at least 64 bits of entropy** to prevent brute-force session guessing attacks. Entropy refers to the amount of randomness or unpredictability in a value. Each "bit" of entropy doubles the number of possible outcomes, meaning a session ID with 64 bits of entropy can have 2<sup>64</sup> possible values. This randomness provides security, as it makes it harder for attackers to guess valid session IDs

2. I updated the content under the "Focus on Hexadecimal Encoding" heading. This paragraph is now:
> Hexadecimal encoding is commonly used in session identifiers. Each hexadecimal character (0-9, A-F) represents 4 bits of entropy. This means that to achieve 64 bits of entropy, a session ID must contain at least 16 random hexadecimal characters, which provides 2<sup>64</sup> possible combinations. When each character is chosen randomly, the session ID offers the desired 64 bits of entropy, making it secure against brute-force guessing attacks. Let’s explore good and bad examples of hexadecimal session IDs to better understand this:

The second update makes the examples slightly redundant, but seeing as this is a source of confusion for some people, I think it's worth being extra clear.

kingthorin reviewed (approved):
Thanks!

kingthorin commented:
Now live: https://owasp.org/www-community/vulnerabilities/Insufficient_Session-ID_Length
